aten/src/ATen/test/tensor_interop_test.cpp

```diff
@@ -0,0 +1,140 @@
+#include "gtest/gtest.h"
+
+#include "ATen/ATen.h"
+#include <caffe2/core/init.h>
+#include <caffe2/core/operator.h>
+
+TEST(TestTensorInterop, Caffe2ToPytorchSimple) {
+  caffe2::Tensor c2_tensor(caffe2::CPU);
+  c2_tensor.Resize(4, 4);
+  auto data = c2_tensor.mutable_data<int64_t>();
+  for (int64_t i = 0; i < 16; i++) {
+    data[i] = i;
+  }
+  at::Tensor at_tensor(c2_tensor.getIntrusivePtr());
+  at::TensorImpl* impl = at_tensor.unsafeGetTensorImpl();
+
+  auto it = impl->data<int64_t>();
+  for (int64_t i = 0; i < 16; i++) {
+    ASSERT_EQ(it[i], i);
+  }
+}
+
+TEST(TestTensorInterop, Caffe2ToPytorchOp) {
+  caffe2::Tensor c2_tensor(caffe2::CPU);
+  c2_tensor.Resize(3, 3);
+  auto data = c2_tensor.mutable_data<int64_t>();
+  for (int64_t i = 0; i < 9; i++) {
+    data[i] = i;
+  }
+  at::Tensor at_tensor(c2_tensor.getIntrusivePtr());
+
+  ASSERT_EQ(at::sum(at_tensor).item<int64_t>(), 36);
+}
+
+TEST(TestTensorInterop, Caffe2ToPytorchUnsupportedDevice) {
+  caffe2::Tensor c2_tensor(caffe2::IDEEP);
+  at::Tensor at_tensor(c2_tensor.getIntrusivePtr());
+  ASSERT_ANY_THROW(at::sum(at_tensor));
+}
+
+TEST(TestTensorInterop, PytorchToCaffe2Op) {
+  caffe2::Workspace workspace;
+  caffe2::NetDef net;
+
+  {
+    auto* blob_a = BlobGetMutableTensor(workspace.CreateBlob("a"), caffe2::CPU);
+    auto at_tensor_a = at::ones({5, 5}, at::dtype(at::kFloat));
+    caffe2::Tensor c2_tensor_a(at_tensor_a.getIntrusivePtr());
+    blob_a->ResizeLike(c2_tensor_a);
+    blob_a->ShareData(c2_tensor_a);
+  }
+
+  {
+    auto* blob_b = BlobGetMutableTensor(workspace.CreateBlob("b"), caffe2::CPU);
+    auto at_tensor_b = at::ones({5, 5}, at::dtype(at::kFloat));
+    caffe2::Tensor c2_tensor_b(at_tensor_b.getIntrusivePtr());
+    blob_b->ResizeLike(c2_tensor_b);
+    blob_b->ShareData(c2_tensor_b);
+  }
+
+  {
+    auto op = net.add_op();
+    op->set_type("Add");
+    op->add_input("a");
+    op->add_input("b");
+    op->add_output("c");
+  }
+
+  workspace.RunNetOnce(net);
+
+  auto result = workspace.GetBlob("c")->GetMutable<caffe2::TensorCPU>();
+  at::TensorImpl* impl = result->unsafeGetTensorImpl();
+
+  auto it = impl->data<float>();
+  for (int64_t i = 0; i < 25; i++) {
+    ASSERT_EQ(it[i], 2.0);
+  }
```

Comment:
There is nothing wrong with this test per se, but I think it tells an incomplete story. This is because I expect `impl->data<float>()` not to do any sort of dispatch. That's fine, and it's important that this works, but we also want to ensure that dispatch works in these cases as well.